File: Window/ShuffleTransformers.m
//
//  ShuffleTransformers.m
//  Cog
//
//  Created by Vincent Spader on 2/27/09.
//  Copyright 2009 __MyCompanyName__. All rights reserved.
//

#import "ShuffleTransformers.h"
@implementation ShuffleImageTransformer

+ (Class)transformedValueClass { return [NSImage class]; }
+ (BOOL)allowsReverseTransformation { return NO; }

// Convert from string to RepeatMode
- (id)transformedValue:(id)value {
	NSLog(@"Transforming value: %@", value);
	
    if (value == nil) return nil;
	
	ShuffleMode mode = [value integerValue];
	
	if (mode == ShuffleOff) {
		return [NSImage imageNamed:@"shuffle_off"];
	}
	else if (mode == ShuffleAlbums) {
		return [NSImage imageNamed:@"shuffle_albums"];
	}
	else if (mode == ShuffleAll) {
		return [NSImage imageNamed:@"shuffle_on"];
	}
	
	return nil;
}

@end


@implementation ShuffleModeTransformer

+ (Class)transformedValueClass { return [NSNumber class]; }
+ (BOOL)allowsReverseTransformation { return YES; }

- (id)initWithMode:(ShuffleMode)s
{
	self = [super init];
	if (self)
	{
		shuffleMode = s;
	}
	
	return self;
}

// Convert from RepeatMode to BOOL
- (id)transformedValue:(id)value {
	NSLog(@"Transforming value: %@", value);
	
    if (value == nil) return nil;
	
	ShuffleMode mode = [value integerValue];
	
	if (shuffleMode == mode) {
		return [NSNumber numberWithBool:YES];
	}
	
	
	return [NSNumber numberWithBool:NO];
}

- (id)reverseTransformedValue:(id)value {
    if (value == nil) return nil;
	
	BOOL enabled = [value boolValue];
	if (enabled) {
		return [NSNumber numberWithInt:shuffleMode];
	}
	else if(shuffleMode == ShuffleOff) {
		return [NSNumber numberWithInt:ShuffleAll];
	}
	else {
		return [NSNumber numberWithInt:ShuffleOff];
	}
}

@end

File: Window/ShuffleTransformers.swift
import Foundation
import AppKit

@objc(ShuffleImageTransformer)
class ShuffleImageTransformer: ValueTransformer {
    override class func transformedValueClass() -> AnyClass {
        return NSImage.self
    }

    override class func allowsReverseTransformation() -> Bool {
        return false
    }

    // Convert from ShuffleMode to NSImage
    override func transformedValue(_ value: Any?) -> Any? {
        guard let value = value else { return nil }

        let mode = ShuffleMode(rawValue: (value as? NSNumber)?.intValue ?? 0) ?? .off
        let shuffleMode = mode as ShuffleMode

        switch shuffleMode {
        case .off:
            return NSImage(named: "shuffleOffTemplate")
        case .albums:
            return NSImage(named: "shuffleAlbumTemplate")
        case .all:
            return NSImage(named: "shuffleOnTemplate")
        default:
            return NSImage(named: "shuffleOffTemplate")
        }
    }
}

@objc(ShuffleModeTransformer)
class ShuffleModeTransformer: ValueTransformer {
    private var shuffleMode: ShuffleMode

    @objc
    init(mode: ShuffleMode) {
        self.shuffleMode = mode
        super.init()
    }

    override class func transformedValueClass() -> AnyClass {
        return NSNumber.self
    }

    override class func allowsReverseTransformation() -> Bool {
        return true
    }

    // Convert from ShuffleMode to BOOL
    override func transformedValue(_ value: Any?) -> Any? {
        guard let value = value else { return nil }

        let mode = ShuffleMode(rawValue: (value as? NSNumber)?.intValue ?? 0) ?? .off

        if shuffleMode == mode {
            return NSNumber(value: true)
        }

        return NSNumber(value: false)
    }

    override func reverseTransformedValue(_ value: Any?) -> Any? {
        guard let value = value else { return nil }

        let enabled = (value as? NSNumber)?.boolValue ?? false
        if enabled {
            return NSNumber(value: shuffleMode.rawValue)
        } else if shuffleMode == .off {
            return NSNumber(value: ShuffleMode.all.rawValue)
        } else {
            return NSNumber(value: ShuffleMode.off.rawValue)
        }
    }
}
